docs/rule/no-action-helper.md

```diff
@@ -0,0 +1,111 @@
+## no-action-helper
+
+What's wrong with `{{action}}`?
+
+"Action" is an overloaded term in Ember parlance. Actions are:
+
+1.) Methods on the `actions` hash
+```js
+actions: {
+  helloWorld() {}
+}
+```
+
+2.) But also modifiers that setup event handlers?
+```hbs
+<div {{action 'helloWorld'}}></div>
+```
+
+3.) Oh, and they are partial applied functions too:
+```hbs
+{{some-component onClick=(action 'addNumbers' 1 2 3)}}
+```
+
+4.) Also, they usually receive strings, but can also receive functions?
+```hbs
+<div {{action this.someMethod}}></div>
+```
+
+They're something we pass downward, but also something we send back up (as in Data Down, Actions Up). They are the way you interact with the DOM, except when you need to actually access the event itself, or when you use the `click` method on a classic component class. The point is, if you try to ask a group of experienced Ember devs to define what an "action" is, you may get a few contradicting/overlapping opinions.
+
+Actions have served many different purposes over the years, and this makes them difficult to learn, to teach, and to repurpose.
+
+```hbs
+<button {{action "increment" 5}}>Click</button>
+<button {{action this.increment 5}}>Click</button>
+<button onclick={{action "increment" 5}}>Click</button>
+<button onclick={{action this.increment 5}}>Click</button>
+<button {{action (action "increment" 5)}}>Click</button>
+<button {{action (action this.increment 5)}}>Click</button>
+```
+
+We wanted to separate out the different responsibilities here, and the `@action` decorator was the first part of the puzzle, and the other two pieces are the `{{fn}}` helper and `{{on}}` modifier.
+
+### Examples
+
+This rule **forbids** the following:
+
+```hbs
+<button {{action 'submit'}}>Submit</button>
+```
+
+```hbs
+<FooBar @baz={{action 'submit'}} />
+```
+
+```hbs
+{{yield (action 'foo')}}
+```
+
+```hbs
+{{yield (action this.foo)}}
+```
+
+This rule **allows** the following:
+
+```hbs
+{{#let (fn this.foo bar) as |action|}}<Component @baz={{action}} />{{/let}}
+```
+
+```hbs
+<MyScope as |action|><Component @baz={{action}} /></MyScope>
```

Comment:
In general, we should not suggest using `action` as a param name like this. Ember is still working on fixing some precedence issues here, and some versions will actually invoke the action helper in this example. It's best to avoid reusing a helper/modifier/component name as a block param name for now...